1. What is Python, and why is it popular


Ans - Python is a programming language.it is a popular language because it is widely used and easy to understand language.  



2. What is an interpreter in Python.

Ans. An interpreter in Python is a program that reads and executes Python code directly, line by line. It translates high-level Python code into machine-level instructions that can be executed by the computer's processor.
Key Points:
It executes code directly, step by step.
If there's a mistake in the code, it stops and shows an error.
It lets you quickly test and see results, especially in its interactive mode, where you can type commands and see outputs instantly.  


3. What are pre-defined keywords in Python

Ans.  Pre-defined keywords in Python are reserved words that have special meanings and purposes in the Python language. These keywords are part of the Python syntax and cannot be used as variable names, function names, or any other identifiers.

1. Case-Sensitive: Keywords must be written in lowercase (e.g., True and true are different).   
2.  Fixed Meaning: The meaning of these keywords is predefined and cannot be changed by the programmer.
3. Integral to Syntax: Keywords are essential for writing Python code and control how the program behaves.

some of the python keywords: and, or, not, if etc.





4. Can keywords be used as variable names

Ans. No, keywords cannot be used as variable names in Python. Keywords are reserved words in Python that have special meanings and are used as part of the language's syntax. Attempting to use a keyword as a variable name will result in a SyntaxError.

# Incorrect: Using a keyword as a variable name
if = 10  # SyntaxError: invalid syntax

# Correct: Using a valid variable name
value = 10

Why Keywords Can't Be Used:
:-Python needs to differentiate between the predefined functionality of keywords and user-defined identifiers like variable names.
:-Keywords are essential for the language's structure, so allowing them as variable names would cause ambiguity.

5. What is mutability in Python
Ans -  Mutability in Python refers to whether an object's value can be changed after it has been created. Based on this property, Python objects are classified into two categories:

Mutable Objects: Their values can be changed after creation.
Immutable Objects: Their values cannot be changed once created.


Mutability in Python refers to whether an object’s value can be changed after it has been created. Based on this property, Python objects are classified into two categories:

Mutable Objects: Their values can be changed after creation.
Immutable Objects: Their values cannot be changed once created.

Mutable Objects
:-Mutable objects allow in-place modifications.
:-Changing a mutable object does not create a new object; the same memory location is updated.
Examples of Mutable Objects:

Lists
Dictionaries
Sets
User-defined objects (typically, unless explicitly designed to be immutable)

6. Why are lists mutable, but tuples are immutable?

ans. The mutability of lists and the immutability of tuples in Python are intentional design choices that serve different purposes and use cases. Here's why they differ:

Why Lists Are Mutable
Purpose:
Lists are designed to be flexible and dynamic. They are commonly used when you need to store and manipulate a collection of items.

Use Cases:

Adding, removing, or modifying elements.
Iterative operations where the collection needs to change.
Implementation:

Lists are implemented with a dynamic array in memory.
This structure allows in-place modifications like changing an element or resizing the list.

Why Tuples Are Immutable
Purpose:
Tuples are meant to represent fixed collections of items. They are often used for grouping related data that should not change, like database records or function return values.

Use Cases:

Ensuring data integrity by preventing accidental modifications.
Using as keys in dictionaries or elements in sets (since immutability is required for hashability).
Implementation:

Tuples are stored more compactly in memory because they do not need to support dynamic resizing.
Immutability simplifies optimization and makes tuples faster in some operations

7. What is the difference between “==” and “is” operators in Python

Ans. The == and is operators in Python are used for comparison, but they serve different purposes:

== (Equality Operator):
Purpose: Compares the values of two objects.
Checks: Whether the data stored in two objects is the same, regardless of whether they are the same object in memory.
Use Case: Use == when you want to compare the contents of two variables.
example
a = [1, 2, 3]
b = [1, 2, 3]

print(a == b)  # Output: True (values are the same)


The == and is operators in Python are used for comparison, but they serve different purposes:

== (Equality Operator):
Purpose: Compares the values of two objects.
Checks: Whether the data stored in two objects is the same, regardless of whether they are the same object in memory.
Use Case: Use == when you want to compare the contents of two variables.
Example:

python
Copy code
a = [1, 2, 3]
b = [1, 2, 3]

is (Identity Operator):
Purpose: Checks whether two references point to the same object in memory.
Checks: Object identity, not value equality.
Use Case: Use is when you want to check if two variables refer to the exact same object.
Conclusion:
Use == when comparing values.
Use is when comparing object identities (memory locations).

8. What are logical operators in Python.

Ans. Logical operators in Python are used to perform logical operations on Boolean values (True or False). They are typically used in conditional statements or loops to combine multiple conditions.

The three main logical operators in Python are:
and:
Returns True if both conditions are True, otherwise returns False.

Example:

a = True
b = False
print(a and b)  # Output: False (since one condition is False)

These logical operators are essential for controlling the flow of execution in Python programs, especially in conditions and loops.

9. What is type casting in Python ?

Ans.  Type casting in Python refers to the process of converting one data type into another. This is often necessary when you want to perform operations that require a specific data type, or when you want to ensure that different types are compatible.

Types of Type Casting:
Implicit Type Casting (Automatic Type Conversion):

Python automatically converts one data type to another when necessary. This usually happens when converting from a smaller to a larger data type (e.g., integer to float).

a = 5       # Integer
b = 2.5     # Float
c = a + b   # Python automatically converts `a` to a float
print(c)    # Output: 7.5

Explicit Type Casting (Manual Type Conversion):

You manually convert a variable from one data type to another using specific functions like int(), float(), str(), etc.
Common Type Casting Functions:

int(): Converts a value to an integer (removes the decimal part for floats).
float(): Converts a value to a floating-point number.



10. What is the difference between implicit and explicit type casting?

Ans. The difference between implicit and explicit type casting in Python lies in how the type conversion is performed and who decides to perform the conversion.

1. Implicit Type Casting (Automatic Type Conversion)
Definition: Implicit type casting happens automatically by Python when it detects that type conversion is necessary.
When: This occurs when you perform operations involving two different data types, and Python automatically converts the smaller type into a larger type (e.g., from integer to float) to avoid data loss.
Who does it: Python itself handles the conversion without any explicit instruction from the programmer.
Examples:
Converting an integer to a float automatically when mixed with floating-point numbers.  

a = 5        # Integer
b = 2.5      # Float
c = a + b    # `a` is automatically converted to float during addition
print(c)     # Output: 7.5

2. Explicit Type Casting (Manual Type Conversion)
Definition: Explicit type casting is when the programmer manually converts one data type to another using built-in functions like int(), float(), str(), etc.
When: This occurs when you want to intentionally convert a value to another data type, and Python does not do it automatically.
Who does it: The programmer explicitly tells Python how to convert the value.
Examples:
Converting a float to an integer manually, losing the decimal part.

a = 3.14      # Float
b = int(a)    # Manually convert float to integer (decimal part is discarded)
print(b)      # Output: 3


11.  What is the purpose of conditional statements in Python?

Ans. Conditional statements in Python allow you to control the flow of a program based on certain conditions. They enable the program to execute specific blocks of code only when certain conditions are true, allowing for decision-making and branching behavior. Conditional statements are essential for making your program dynamic and responsive to different inputs and situations.

Purpose of Conditional Statements:
Control Program Flow: Conditional statements help decide which part of the code to execute based on the evaluation of conditions.
Make Decisions: They allow the program to choose between different options based on logic (e.g., "If this is true, do this; otherwise, do that").
Handle Multiple Scenarios: You can use conditionals to handle different scenarios and behaviors depending on the values of variables or inputs.
Improve Program Logic: They enable you to implement logic such as comparing values, checking for specific conditions, and running appropriate code based on those checks.
Types of Conditional Statements in Python:
if statement:

The if statement is used to check if a condition is True. If the condition is True, the block of code under it is executed.
Example:

python
Copy code

age = 18
if age >= 18:
    print("You are an adult.")
    
Conditional statements are crucial for implementing decision-making and controlling program flow in Python. They let you define rules and behavior that vary depending on specific conditions or inputs. Without them, your program would lack flexibility and could not adapt to different scenarios.









12. How does the elif statement work.

Ans. The elif (short for else if) statement in Python is used to check multiple conditions in a sequence. It allows you to test several conditions in a sequence, executing the corresponding block of code for the first condition that evaluates to True. If none of the if or elif conditions are True, the final else block is executed (if provided).

How elif Works:
The if statement is evaluated first. If the condition is True, the code inside the if block is executed, and the rest of the elif and else blocks are skipped.
If the if condition is False, Python checks the conditions in the elif statements in the order they appear.
The first elif condition that evaluates to True will execute its associated block of code, and the remaining conditions are ignored.
If none of the if or elif conditions are True, the code inside the else block (if provided) is executed.
Syntax of elif:

day = "Monday"

if day == "Monday":
    print("Start of the week.")
elif day == "Tuesday":
    print("Second day of the week.")
elif day == "Wednesday":
    print("Mid-week.")
elif day == "Friday":
    print("Almost weekend.")
else:
    print("Weekend or another day.")

Summary:
elif allows you to check multiple conditions after the initial if statement.
It makes decision-making more efficient when you have several distinct conditions to evaluate.







13. What is the difference between for and while loops ?

Ans. The for and while loops in Python are both used to repeat a block of code multiple times, but they differ in how they control the loop and when they terminate.

1. for Loop:
Purpose: Used to iterate over a sequence (like a list, tuple, string, or range).
How It Works: The loop executes a block of code for each item in the sequence.
Termination: The loop terminates once it has iterated over all elements in the sequence.
Syntax:

fruits = ["apple", "banana", "cherry"]
for fruit in fruits:
    print(fruit)

Outcome

apple
banana
cherry


The for and while loops in Python are both used to repeat a block of code multiple times, but they differ in how they control the loop and when they terminate.

1. for Loop:
Purpose: Used to iterate over a sequence (like a list, tuple, string, or range).
How It Works: The loop executes a block of code for each item in the sequence.
Termination: The loop terminates once it has iterated over all elements in the sequence.
Syntax:
python
Copy code
for item in sequence:
    # Code to execute for each item in the sequence
Example:
python
Copy code
fruits = ["apple", "banana", "cherry"]
for fruit in fruits:
    print(fruit)
Output:

Copy code
apple
banana
cherry

Use Case: When you know the number of iterations beforehand or you are iterating over a known sequence (like a list or range).
2. while Loop:
Purpose: Used to repeat a block of code as long as a specified condition is True.
How It Works: The loop continues to execute the block of code until the condition becomes False. If the condition is False initially, the code inside the loop may never execute.
Termination: The loop will terminate as soon as the condition becomes False.
Syntax:

Example:
count = 1
while count <= 3:
    print(count)
    count += 1

output
    
    1
    2
    3

Use Case: When the number of iterations is unknown beforehand, and you want to repeat the loop until a specific condition is met (like waiting for a user input or a dynamic condition).






14. Describe a scenario where a while loop is more suitable than a for loop.

Ans. A while loop is more suitable than a for loop in scenarios where you don’t know the exact number of iterations in advance, and the loop needs to continue running until a certain condition is met. This makes it ideal for situations where the loop is based on dynamic or user-defined conditions that can change during the execution of the loop.

Scenario: Asking for Valid User Input
Let’s say you’re creating a program that asks a user to input their age, and you want to ensure that the input is a valid number (greater than 0). If the user provides invalid input, the program should continue asking until a valid age is entered. This is a situation where a while loop is more suitable than a for loop because you don't know how many times the user will provide invalid input, and you only want to stop when a valid response is received.

Code Example:

age = -1  # Initialize to an invalid age

while age <= 0:
    try:
        age = int(input("Please enter your age: "))
        if age <= 0:
            print("Age must be a positive number. Please try again.")
    except ValueError:
        print("Invalid input! Please enter a valid number.")
    
print(f"Thank you! Your age is {age}.")


Explanation:
The while loop runs as long as the condition age <= 0 is True, meaning it will keep asking for the user's age until a valid positive number is provided.
The loop checks if the entered age is greater than 0. If not, it will continue prompting the user for a valid input.
The loop doesn't know how many iterations it will take because the number of attempts depends on when the user provides a valid response.
Why a while Loop Is Suitable Here:
Uncertain Number of Iterations: You don’t know in advance how many times the user might give an invalid input. A while loop will continue asking for input until a valid value is provided.
Dynamic Condition: The condition (age <= 0) changes based on user input, making a while loop the ideal choice because the loop will run until the condition is no longer True.
This scenario is an excellent example of when a while loop is more suitable than a for loop, as the number of iterations is not known and is dependent on external input.








# #Practical Questions

1. Write a Python program to print "Hello, World!_

Ans. in the following code


In [3]:
print("Hello, World!-")

Hello, World!-


2. Write a Python program that displays your name and age?

Ans.  check in the following code.


In [5]:
# Displaying name and age
name = "Rahul Joshi"
age = 24

print("Name:", name)
print("Age:", age)


Name: Rahul Joshi
Age: 24


3. Write code to print all the pre-defined keywords in Python using the keyword library.


In [6]:
import keyword

# Print all the pre-defined keywords in Python
print("List of Python keywords:")
print(keyword.kwlist)


List of Python keywords:
['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


4. Write a program that checks if a given word is a Python keyword


In [7]:
import keyword

# Function to check if a word is a Python keyword
def is_python_keyword(word):
    if keyword.iskeyword(word):
        return True
    else:
        return False

# Input from user
word = input("Enter a word to check if it's a Python keyword: ")

# Check and display result
if is_python_keyword(word):
    print(f"'{word}' is a Python keyword.")
else:
    print(f"'{word}' is NOT a Python keyword.")


Enter a word to check if it's a Python keyword: if
'if' is a Python keyword.


5. Create a list and tuple in Python, and demonstrate how attempting to change an element works differently
for each.

Here's an example demonstrating how you can create a list and a tuple in Python, and how attempting to change an element works differently for each:

In [8]:
# Creating a list
my_list = [10, 20, 30, 40]

# Attempting to change an element in the list
print("Original list:", my_list)
my_list[1] = 100  # Changing the second element
print("Updated list:", my_list)

# Creating a tuple
my_tuple = (10, 20, 30, 40)

# Attempting to change an element in the tuple
print("\nOriginal tuple:", my_tuple)
try:
    my_tuple[1] = 100  # Trying to change the second element
except TypeError as e:
    print("Error while trying to modify tuple:", e)


Original list: [10, 20, 30, 40]
Updated list: [10, 100, 30, 40]

Original tuple: (10, 20, 30, 40)
Error while trying to modify tuple: 'tuple' object does not support item assignment


Explanation:
List:

my_list = [10, 20, 30, 40]: A list is created with four elements.
Lists are mutable, which means you can change or modify elements. In this case, my_list[1] = 100 successfully changes the second element from 20 to 100.
Tuple:

my_tuple = (10, 20, 30, 40): A tuple is created with four elements.
Tuples are immutable, which means you cannot modify their elements after creation. Trying to execute my_tuple[1] = 100 will raise a TypeError because tuples cannot be changed once created.

Key Points:
Lists are mutable, so their elements can be modified after creation.
Tuples are immutable, so you cannot change their elements once they are created.







6. Write a function to demonstrate the behavior of mutable and immutable arguments.

In Python, mutable objects (like lists, dictionaries, and sets) can be modified inside a function, whereas immutable objects (like integers, strings, and tuples) cannot be modified inside a function. Here's a function that demonstrates the behavior of mutable and immutable arguments:

Example:

In [9]:
def demonstrate_mutable_immutable(mutable_arg, immutable_arg):
    # Modifying mutable argument inside the function
    mutable_arg.append(4)  # Adds 4 to the list

    # Trying to modify immutable argument inside the function
    immutable_arg = 100  # This will reassign the variable, not modify the original

    # Printing the results
    print("Inside function:")
    print("Modified mutable argument:", mutable_arg)  # List should be modified
    print("Modified immutable argument:", immutable_arg)  # This will show 100, but it doesn't change the original

# Test with mutable and immutable arguments
my_list = [1, 2, 3]  # Mutable
my_number = 10  # Immutable

# Calling the function
print("Before function call:")
print("Original mutable argument:", my_list)
print("Original immutable argument:", my_number)

demonstrate_mutable_immutable(my_list, my_number)

print("\nAfter function call:")
print("Original mutable argument:", my_list)  # Should be modified
print("Original immutable argument:", my_number)  # Should remain unchanged


Before function call:
Original mutable argument: [1, 2, 3]
Original immutable argument: 10
Inside function:
Modified mutable argument: [1, 2, 3, 4]
Modified immutable argument: 100

After function call:
Original mutable argument: [1, 2, 3, 4]
Original immutable argument: 10


Explanation:

Mutable Argument (my_list):

We pass my_list (which is mutable) into the function. Inside the function, the list is modified by appending an element to it (mutable_arg.append(4)).
Since lists are mutable, the change is reflected outside the function as well.

Immutable Argument (my_number):

We pass my_number (which is immutable) into the function. Inside the function, we attempt to reassign the value of immutable_arg = 100. However, this creates a new local reference to a different integer value. The original value of my_number is not modified.

Since integers are immutable, reassigning the value doesn't affect the original variable outside the function.

Key Points:
Mutable objects (like lists) can be modified inside a function, and the changes will persist outside the function.
Immutable objects (like integers) cannot be modified inside a function. Reassigning them creates a new object, and the original object remains unchanged outside the function.









7. Write a function to demonstrate the behavior of mutable and immutable arguments.

Ans. Here's a Python function that demonstrates the behavior of mutable and immutable arguments:

In [10]:
def demonstrate_mutable_immutable(mutable_arg, immutable_arg):
    # Modifying mutable argument inside the function
    print("Inside the function:")
    print(f"Original mutable_arg before modification: {mutable_arg}")
    mutable_arg.append(10)  # Modify the mutable argument (list)
    print(f"mutable_arg after modification: {mutable_arg}")

    # Trying to modify immutable argument inside the function
    print(f"Original immutable_arg before reassignment: {immutable_arg}")
    immutable_arg = 20  # Reassigning the immutable argument (int)
    print(f"immutable_arg after reassignment: {immutable_arg}")

    print("\nEnd of function.\n")

# Test with mutable and immutable arguments
my_list = [1, 2, 3]  # Mutable (list)
my_number = 5  # Immutable (integer)

# Call the function
print("Before function call:")
print(f"Original mutable argument (list): {my_list}")
print(f"Original immutable argument (number): {my_number}")

demonstrate_mutable_immutable(my_list, my_number)

print("After function call:")
print(f"Original mutable argument (list) after function call: {my_list}")
print(f"Original immutable argument (number) after function call: {my_number}")


Before function call:
Original mutable argument (list): [1, 2, 3]
Original immutable argument (number): 5
Inside the function:
Original mutable_arg before modification: [1, 2, 3]
mutable_arg after modification: [1, 2, 3, 10]
Original immutable_arg before reassignment: 5
immutable_arg after reassignment: 20

End of function.

After function call:
Original mutable argument (list) after function call: [1, 2, 3, 10]
Original immutable argument (number) after function call: 5


Explanation:
Mutable Argument (my_list):

The function demonstrate_mutable_immutable() takes two arguments: mutable_arg and immutable_arg.
Inside the function, we modify mutable_arg, which is a list (my_list). The method .append() is used to add an element to the list. Since lists are mutable, this modification affects the original my_list outside the function as well.
Immutable Argument (my_number):

The variable immutable_arg is passed as an integer (my_number), which is immutable. Inside the function, we attempt to reassign immutable_arg = 20, which does not modify the original my_number. Instead, it creates a new local reference, leaving the original my_number unchanged outside the function.

Key Points:
Mutable objects (like lists) can be changed inside a function, and the change is reflected outside the function.
Immutable objects (like integers) cannot be modified inside a function. Reassigning the value inside the function only affects the local copy, and the original value remains unchanged outside the function.

8. Write a program to demonstrate the use of logical operators


Ans. Logical operators in Python are used to combine multiple conditions or expressions. The three main logical operators are:

and: Returns True if both conditions are True.
or: Returns True if at least one condition is True.
not: Reverses the logical state, returns True if the condition is False, and False if the condition is True.
Here is a Python program that demonstrates the use of logical operators:

Code:


In [11]:
# Demonstrating logical operators in Python

# Initial values for the variables
a = 10
b = 20
c = 30

# Using 'and' operator
if a < b and b < c:
    print("Both conditions are True for 'and' operator.")
else:
    print("One or both conditions are False for 'and' operator.")

# Using 'or' operator
if a > b or b < c:
    print("At least one condition is True for 'or' operator.")
else:
    print("Both conditions are False for 'or' operator.")

# Using 'not' operator
if not a > b:
    print("Condition is False, so 'not' reverses it to True.")
else:
    print("Condition is True, so 'not' reverses it to False.")


Both conditions are True for 'and' operator.
At least one condition is True for 'or' operator.
Condition is False, so 'not' reverses it to True.


Explanation:
and Operator:

The condition a < b and b < c is checked. Both conditions must be True for the overall condition to be True.
Since a is less than b and b is less than c, the output will indicate that both conditions are true for the and operator.
or Operator:

The condition a > b or b < c is checked. The overall condition will be True if at least one of the conditions is True.
Here, a > b is False, but b < c is True, so the output will indicate that at least one condition is true for the or operator.
not Operator:

The condition not a > b is checked. The not operator reverses the condition.
Since a > b is False, applying not makes it True, and the output will reflect that the condition is reversed by not.

Key Points:
The and operator returns True only if both conditions are True.
The or operator returns True if at least one condition is True.
The not operator inverts the truth value of a condition. If the condition is True, it returns False, and if it's False, it returns True.

9.  Write a Python program to convert user input from string to integer, float, and boolean types.

Ans - Here's a Python program that demonstrates how to convert user input from a string to different data types, including integer, float, and boolean:

Code:


In [12]:
# Get user input as a string
user_input = input("Enter a value: ")

# Convert string to integer
try:
    user_input_int = int(user_input)
    print(f"The input as an integer: {user_input_int}")
except ValueError:
    print("The input is not a valid integer.")

# Convert string to float
try:
    user_input_float = float(user_input)
    print(f"The input as a float: {user_input_float}")
except ValueError:
    print("The input is not a valid float.")

# Convert string to boolean
if user_input.lower() in ['true', '1', 't', 'y', 'yes']:
    user_input_bool = True
elif user_input.lower() in ['false', '0', 'f', 'n', 'no']:
    user_input_bool = False
else:
    print("Invalid input for boolean conversion.")
    user_input_bool = None

if user_input_bool is not None:
    print(f"The input as a boolean: {user_input_bool}")


Enter a value: 123
The input as an integer: 123
The input as a float: 123.0
Invalid input for boolean conversion.


Key Points:
The input() function always returns a string. You need to explicitly convert it to the desired data type (integer, float, or boolean).
The program uses exception handling (try-except) to manage invalid inputs for integer and float conversions.
Boolean conversion checks if the string is a common representation of True or False.








10. Write code to demonstrate type casting with list elements.

Ans. Here is a Python program that demonstrates type casting with elements in a list. In this example, we will convert elements in a list from one data type to another:

Code:

In [13]:
# Original list with mixed data types (string numbers, integer, and float)
original_list = ["10", "20.5", "30", 40, 50.75]

# Convert all elements to integers
int_list = [int(float(item)) if isinstance(item, str) else int(item) for item in original_list]
print("List after converting all elements to integers:", int_list)

# Convert all elements to floats
float_list = [float(item) if isinstance(item, str) else float(item) for item in original_list]
print("List after converting all elements to floats:", float_list)

# Convert all elements to strings
string_list = [str(item) for item in original_list]
print("List after converting all elements to strings:", string_list)

# Convert all elements to booleans
boolean_list = [bool(item) for item in original_list]
print("List after converting all elements to booleans:", boolean_list)


List after converting all elements to integers: [10, 20, 30, 40, 50]
List after converting all elements to floats: [10.0, 20.5, 30.0, 40.0, 50.75]
List after converting all elements to strings: ['10', '20.5', '30', '40', '50.75']
List after converting all elements to booleans: [True, True, True, True, True]


Explanation:
We start with a list original_list that contains elements of different types: some are strings representing numbers, and others are integers or floats.

Convert to Integers:

Each element is first converted to a float (if it is a string representing a number) and then to an int.
For other types (integers or floats), it directly converts to an integer.
Convert to Floats:

Each element is converted to a float. If the element is a string, it's first converted to a float. If it's already an integer or float, it's directly cast to a float.
Convert to Strings:

Each element is converted to a string using the str() function.
Convert to Booleans:

The bool() function converts each element to a boolean value.
For Python, any non-zero number (including non-empty strings or non-empty lists) evaluates to True, and 0 or an empty string/list evaluates to False.


Key Points:
Type casting is used to convert data from one type to another, which is especially useful when dealing with mixed data types in a list.
The list comprehension technique is used to apply type casting to each element of the list.
The isinstance() function is used to check the type of each element before casting it (e.g., checking if the element is a string before trying to convert it to a float.

11. Write a program that checks if a number is positive, negative, or zero.

Ans. Here is a simple Python program that checks if a given number is positive, negative, or zero:

Code:

In [14]:
# Get user input
number = float(input("Enter a number: "))

# Check if the number is positive, negative, or zero
if number > 0:
    print("The number is positive.")
elif number < 0:
    print("The number is negative.")
else:
    print("The number is zero.")


Enter a number: 2
The number is positive.


Explanation:
Input: The program uses input() to get the number from the user. The input is then converted to a float to handle both integer and floating-point numbers.

Condition Check:

Positive: If the number is greater than 0, the program will print "The number is positive."
Negative: If the number is less than 0, the program will print "The number is negative."
Zero: If the number is equal to 0, the program will print "The number is zero."

Key Points:
The program handles both integers and floating-point numbers by converting the input to a float.
It uses a simple series of if, elif, and else statements to check if the number is positive, negative, or zero.

12. Write a for loop to print numbers from 1 to 100

Ans. Here is a Python program using a for loop to print numbers from 1 to 100:

Code:

In [15]:
# Using a for loop to print numbers from 1 to 100
for num in range(1, 101):
    print(num)


1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100


Key Points:
The range() function is very useful when you need to generate a sequence of numbers.
The for loop will iterate through each number in the range and print it one by one.







13. Write a Python program to find the sum of all even numbers between 1 and 50

Ans. Here's a Python program that calculates the sum of all even numbers between 1 and 50:

Code:

In [16]:
# Initialize sum variable
sum_even = 0

# Loop through numbers from 1 to 50
for num in range(1, 51):
    if num % 2 == 0:  # Check if the number is even
        sum_even += num  # Add the even number to the sum

# Print the result
print("The sum of all even numbers between 1 and 50 is:", sum_even)


The sum of all even numbers between 1 and 50 is: 650


Explanation:
range(1, 51): Generates numbers from 1 to 50 (inclusive).
if num % 2 == 0: This condition checks if the number is even. If the remainder when dividing num by 2 is 0, the number is even.
sum_even += num: If the number is even, it is added to the variable sum_even.
Result: After the loop finishes, the program prints the sum of all even numbers between 1 and 50.

Key Points:
The % (modulus) operator is used to check if a number is divisible by 2.
This program efficiently sums the even numbers by looping only once through the range.







14. Write a program to reverse a string using a while loop

Ans. Here's a Python program that reverses a string using a while loop:

Code:

In [17]:
# Input string from the user
input_string = input("Enter a string: ")

# Initialize an empty string to store the reversed string
reversed_string = ""

# Set an index variable to point to the last character of the string
index = len(input_string) - 1

# Use a while loop to iterate through the string from the last character to the first
while index >= 0:
    reversed_string += input_string[index]  # Add the current character to the reversed string
    index -= 1  # Move to the previous character

# Print the reversed string
print("The reversed string is:", reversed_string)


Enter a string: Rahul
The reversed string is: luhaR


Key Points:
The while loop helps iterate through the string backwards by decreasing the index.
The += operator is used to build the reversed string step by step.

15. Write a Python program to calculate the factorial of a number provided by the user using a while loop

Ans. Here is a Python program that calculates the factorial of a number provided by the user using a while loop:

Code:

In [18]:
# Input number from the user
number = int(input("Enter a number to find its factorial: "))

# Initialize the factorial result to 1
factorial = 1

# Check if the number is negative, as factorials are not defined for negative numbers
if number < 0:
    print("Factorial is not defined for negative numbers.")
else:
    # Use a while loop to calculate the factorial
    while number > 0:
        factorial *= number  # Multiply the current factorial by the number
        number -= 1  # Decrement the number by 1

    # Print the result
    print("The factorial is:", factorial)


Enter a number to find its factorial: 5
The factorial is: 120


Key Points:
Factorial of a number n is the product of all positive integers less than or equal to n. For example, 5! = 5 * 4 * 3 * 2 * 1 = 120.
The while loop keeps multiplying the current value of factorial with the current number and then reduces the number until it becomes 0.



# thanks this is my first assignment made by Rahul Joshi